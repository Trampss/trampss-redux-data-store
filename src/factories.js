import * as types from './types/index'
import reducer from './reducer'

const defaultOptions = {
  key: 'id',
  type: 'keyValue',
  prefix: '',
}

const simpleDefaultData = {
  bool: false,
  string: '',
  array: [],
  object: {},
  number: 0,
}

const getWrappedStore = (middlewares = {}) => (options = {}) => {
  const innerOptions = { ...defaultOptions, ...options }
  const {
    key,
    type,
    prefix,
    name,
    defaultData,
  } = innerOptions

  // eslint-disable-next-line max-len
  if (type === 'simpleObject') console.warn('/k-redux-factory/ You are using a deprecated "simpleObject" type. We recommend using one of these types: simple.object, simple.array, simple.bool or simple.string.')

  const [innerType, subType] = type.split('.')
  const typeConfig = types[innerType]

  // get default data
  // - default is the given one
  // - if there is no given one we try to retrieve the one associated with the type (simple.<type>)
  let innerDefaultData
  if (innerType === 'simple' || type === 'simpleObject') {
    innerDefaultData = defaultData
    if (innerDefaultData === undefined && subType) {
      innerDefaultData = simpleDefaultData[subType]
    }
  }

  return Object.assign(
    reducer({ ...middlewares, engine: typeConfig.middlewares })(key)(prefix)(name)(innerDefaultData),

    // type (debug purpose)
    { krfType: type },

    // actions
    ...Object.keys(typeConfig.actions).map(k => ({ [k]: typeConfig.actions[k](prefix)(name) })),

    // selectors
    ...typeConfig.selectors(innerOptions),
  )
}

// error :( - not a middleware nor an option parameter
const error = () => { throw Error('parameter is not a middleware configuration, nor a factory option object.') }

// params checkers
const isMiddleware = params => params.engine || params.pre || params.post
const isOptions = params => (!!params.name) || (typeof params === 'string')

export const factory = (forcedOptions = {}) => (params) => {
  // no param : error
  if (params === null || params === undefined) return error()

  // middleware case
  if (isMiddleware(params)) {
    return options => getWrappedStore(params)({ ...options, ...forcedOptions })
  }
  // no middleware case
  if (isOptions(params)) {
    if (typeof params === 'string') return getWrappedStore()({ name: params, ...forcedOptions })
    return getWrappedStore()({ ...params, ...forcedOptions })
  }

  // not a valid param
  return error()
}

export const keyValue = factory({ type: 'keyValue' })

export const simple = factory({ type: 'simple' })
Object.assign(
  simple,
  {
    object: factory({ type: 'simple.object' }),
    bool: factory({ type: 'simple.bool' }),
    string: factory({ type: 'simple.string' }),
    array: factory({ type: 'simple.array' }),
    number: factory({ type: 'simple.number' }),
  },
)

// older method deprecated
export const simpleObject = simple
