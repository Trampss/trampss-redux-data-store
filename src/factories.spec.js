/* eslint-env jest */
import { simple, keyValue } from './factories'

describe('factories', () => {
  it('should use options of object type', () => expect(simple.object('object').options).toMatchSnapshot())
  it('should use options of specific string type', () => expect(simple.string('string').options).toMatchSnapshot())
  it('should use options of specific array type', () => expect(simple.array('array').options).toMatchSnapshot())
  it('should use options of specific boolean type', () => expect(simple.bool('boolean').options).toMatchSnapshot())
  it('should use options of specific keyValue type', () => expect(keyValue('keyValue').options).toMatchSnapshot())

  it('should use overwrited options of object type', () => expect(simple.object({ name: 'object', defaultData: { o: 'test' } }).options).toMatchSnapshot())
  it('should use overwrited options of specific string type', () => expect(simple.string({ name: 'string', defaultData: 'test' }).options).toMatchSnapshot())
  it('should use overwrited options of specific array type', () => expect(simple.array({ name: 'array', defaultData: ['test', 1, 2] }).options).toMatchSnapshot())
  it('should use overwrited options of specific boolean type', () => expect(simple.bool({ name: 'boolean', defaultData: true }).options).toMatchSnapshot())
  it('should use overwrited options of specific keyValue type', () => expect(keyValue({
    name: 'string',
    defaultData: [
      { id: 1, value: 'test1' },
      { id: 2, value: 'test2' },
    ],
  }).options).toMatchSnapshot())
})
