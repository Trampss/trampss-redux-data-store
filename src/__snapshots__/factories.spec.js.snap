// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factories should use options of object type 1`] = `
Object {
  "defaultData": Object {},
  "key": "id",
  "name": "object",
  "prefix": "",
  "type": "simple.object",
}
`;

exports[`factories should use options of specific array type 1`] = `
Object {
  "defaultData": Array [],
  "key": "id",
  "name": "array",
  "prefix": "",
  "type": "simple.array",
}
`;

exports[`factories should use options of specific boolean type 1`] = `
Object {
  "defaultData": false,
  "key": "id",
  "name": "boolean",
  "prefix": "",
  "type": "simple.bool",
}
`;

exports[`factories should use options of specific keyValue type 1`] = `
Object {
  "key": "id",
  "name": "keyValue",
  "prefix": "",
  "type": "keyValue",
}
`;

exports[`factories should use options of specific string type 1`] = `
Object {
  "defaultData": "",
  "key": "id",
  "name": "string",
  "prefix": "",
  "type": "simple.string",
}
`;

exports[`factories should use overwrited options of object type 1`] = `
Object {
  "defaultData": Object {
    "o": "test",
  },
  "key": "id",
  "name": "object",
  "prefix": "",
  "type": "simple.object",
}
`;

exports[`factories should use overwrited options of specific array type 1`] = `
Object {
  "defaultData": Array [
    "test",
    1,
    2,
  ],
  "key": "id",
  "name": "array",
  "prefix": "",
  "type": "simple.array",
}
`;

exports[`factories should use overwrited options of specific boolean type 1`] = `
Object {
  "defaultData": true,
  "key": "id",
  "name": "boolean",
  "prefix": "",
  "type": "simple.bool",
}
`;

exports[`factories should use overwrited options of specific keyValue type 1`] = `
Object {
  "defaultData": Array [
    Object {
      "id": 1,
      "value": "test1",
    },
    Object {
      "id": 2,
      "value": "test2",
    },
  ],
  "key": "id",
  "name": "string",
  "prefix": "",
  "type": "keyValue",
}
`;

exports[`factories should use overwrited options of specific string type 1`] = `
Object {
  "defaultData": "test",
  "key": "id",
  "name": "string",
  "prefix": "",
  "type": "simple.string",
}
`;
